### ISSUE TITLE
Complexity from closures is counted in method complexity

### ISSUE THREAD

**Issue by TomBZombie** [phpmd/phpmd#296]
Consider the following code:

```php

public function foo() {
	for ($i = 0; $i < 100; $i++) {

	}
	
	$x = function() {
		if ($whatever) {

		}
	};
        return $x;
}

```

The NPath complexity and Cyclomatic Complexity are both calculated as both the foo function's complexity and the anonymous function's complexity. For the purposes of any statistical analysis these should be considered as separate functions. Other tools such as http://jsmeter.info/ correctly handle closures/anonymous functions.

**Comment by kylekatarnls** [comment]
I don't think so, you're case is an exception because it returns 1 closure so we may exclude this closure. But in many case, this would just hide actual complexity:

With the following code:
```php
public function foo() {
    for ($i = 0; $i < 100; $i++) {
        if ($i % 2) {
            echo '.';
        }
    }
}
```

Changing it using closure this way:
```php
public function foo() {
    $echo = function ($i) {
        if ($i % 2) {
            echo '.';
        }
    };

    for ($i = 0; $i < 100; $i++) {
        $echo($i);
    }
}
```

Obviously, this does not reduce the complexity of `foo()` so you should get at least an as big NPath.

The only right way to reduce complexity is to use non-anonymous functions:
```php
private function echo($i) {
    if ($i % 2) {
        echo '.';
    }
}

public function foo() {
    for ($i = 0; $i < 100; $i++) {
        $this->echo($i);
    }
}
```

I vote for not supporting this change.

**Comment by TRPB** [comment]
Most of the time anonymous functions are used, they are passed into a different method to be called. Your example above would be in the minority, at least where I've seen anonymous functions used. Defining them and calling them in the same location rather defeats the purpose of having them.

Consider:
```
array_map(function($array) { 
//...
}, $array);
```

It can also be expressed as 

```
array_map([$this, 'funcName'], $array);
```

but is often less clear. The result is identical and the logic separated in the same way. The anonymous function is a separate code block and should be counted separately. As I noted in the original issue, in JavaScript, where anonymous functions are everywhere, they are counted as their own functions in metrics.

**Comment by kylekatarnls** [comment]
"less clear"? You find anonymous function more clear than named methods?

Returning a closure could be an example of good exception, but your example is exactly the kind of thing the complexity count try to help you to avoid.

If the complexity is counted by method, it's because a method/function is an atomic process you can test on its own, you can give it a name that explain what it does and a PHPDoc that gives even more details about this enclosed process, input and output.

When using a closure in things like `array_map`, you can't test this closure alone, and this function has no name by definition.

See those both:
```php
$list = [];

foreach ($input as $item) {
  $list[] = $item % 1 ? 'odd' : 'even';
}
```
And
```php
$list = array_map(function ($item) {
  return $item % 1 ? 'odd' : 'even';
}, $input);
```

They are obviously the same algorithm, and logically the same complexity, you do not reduce complexity by using array_map over foreach, if you use a tool that tells you it does, sorry but it's wrong, you're just hiding complexity under the carpet.

By doing:
```php
/**
 * Get "odd" or "even" according to given number.
 */
function getOddOrEven($item) {
  return $item % 1 ? 'odd' : 'even';
}

$list = array_map('getOddOrEven', $input);
```
Here you really reduce complexity by creating a pure reusable and documented function dedicated to an atomic purpose, then calling it in an other function/method that did not take a responsibility it can delegate.

This way, you better respect both Single-responsibility Principle (dedicated function) and Open-closed Principle from the SOLID best-practices.

And it's certainly not less clear. I can't agree your point on this example.

**Comment by kylekatarnls** [comment]
An other example that makes it clear closures do not reduce complexity is nesting:

```php
function parseItems($items) {

return array_map(function ($item) {
  if (has_stuff($item)) {
    do_stuff($item);
  }

  return array_map(function ($subItem) {
    if (has_other_stuff($subItem)) {
      do_other_stuff($subItem);
    }
      return array_map(function ($subSubItem) {
        if (has_other_stuff($subSubItem)) {
          do_other_stuff($subSubItem);
        }

        return $subSubItem;
    }, $subItem);
  }, $item);
}, $items);

}
```
Are you sure parseItems complexity should be `1`? I don't think so.

**Comment by kylekatarnls** [comment]
I see ESLint and JSHint does not count it this way. I don't see the point of using complexity metric if replacing foreach with array_map (or for..in with .map in JS) make your function magically fall under the expected maximum. For me it's completely artificial.

The following:
```js
function foo(items) {
  if (!items) {
    console.log('no items');
  }

  if (items) {
    console.log('items');
  }
}
```
gives me `3`, while:
```js
function foo(items) {
  if (!items) {
    console.log('no items');
  }

  (function () {
    if (items) {
      console.log('items');
    }
  })();
}
```
gives me `2`. It's absurd, I did not reduce the complexity by encapsulating the second statement in an anonymous function. And it's not the [definition of the metric](https://en.wikipedia.org/wiki/Cyclomatic_complexity)

I can't tell you why those tools count it this way, it does not makes sense to me. But as for as PHPMD is concerned regarding its purpose to improve the code, it would be more harmful than helpful in the majority of cases.

**Comment by jakzal** [comment]
I agree with @kylekatarnls here. Complexity of a method is calculated based on its method's body. An anonymous function declared inside a method's body is indeed part of it.

Otherwise the anonymous' function complexity wouldn't be accounted for anywhere. You'd be able to game the system to get CC of 1 by enclosing complex methods in anonymous functions.
